### ISSUE TITLE
Fiber.park is abnormal when call Fiber.unpark immediately after Fiber.start.

### ISSUE THREAD

**Issue by thelastender** [puniverse/quasar#247]
I have encountered a abnormal behavior when I call Fiber.unpark immediately after Fiber.start .
Here is the example:

```
public class MyFiberTest
{
	public static void main(String[] args)
	{
		CountDownLatch countDownLatch = new CountDownLatch(1);
		FiberScheduler fiberScheduler = new FiberExecutorScheduler("system", Executors.newWorkStealingPool());
		B b = new B();
		Fiber<Object> f = new Fiber<>(fiberScheduler, new SuspendableRunnable()
		{
			@Override
			public void run() throws SuspendExecution, InterruptedException
			{

				for(int i = 0; i < 10; ++i)
				{
					LoggerFactory.getLogger(this.getClass()).debug("before {}", i);
					b.f();
					LoggerFactory.getLogger(this.getClass()).debug("after {}", i);
				}
			}
		});
		f.start();
		f.unpark();
		try
		{
			countDownLatch.await();
		}
		catch(InterruptedException e)
		{
			e.printStackTrace();
		}
	}

	public static class B
	{
		@Suspendable
		public void f()
		{
			System.out.println("f in B");
			try
			{
				g();
			}
			catch(SuspendExecution suspendExecution)
			{
				throw new AssertionError(suspendExecution);
			}
		}

		private void g() throws SuspendExecution
		{
			Fiber.park(3, TimeUnit.SECONDS);
		}
	}
}
```

The output is:

```
2017-03-03 14:48:41.240 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 0
f in B
2017-03-03 14:48:41.247 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 0
2017-03-03 14:48:41.247 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 1
f in B
```

Wait 3 seconds then print:

```
2017-03-03 14:48:44.248 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 1
2017-03-03 14:48:44.248 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 2
f in B
2017-03-03 14:48:44.248 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 2
2017-03-03 14:48:44.249 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 3
f in B
```

Wait 3 seconds then print:

```
2017-03-03 14:48:47.249 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 3
2017-03-03 14:48:47.249 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 4
f in B
2017-03-03 14:48:47.249 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 4
2017-03-03 14:48:47.250 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 5
f in B
```

Wait 3 seconds then print:

```
2017-03-03 14:48:50.250 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 5
2017-03-03 14:48:50.250 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 6
f in B
2017-03-03 14:48:53.254 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 6
2017-03-03 14:48:53.254 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 7
f in B
```

Wait 3 seconds then print:

```
2017-03-03 14:48:56.255 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 7
2017-03-03 14:48:56.255 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 8
f in B
2017-03-03 14:48:59.258 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 8
2017-03-03 14:48:59.258 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 9
f in B
```

Wait 3 seconds then print:

```
2017-03-03 14:49:02.259 DEBUG                                 [@] (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 9
```

This shows that the Fiber.park is not to block between odd number. 

I think this may be a bug? or is there a rule that can not call Fiber.park immediately after a Fiber.start . If so, I think this rule is too strange and may cause many bugs.

**Comment by thelastender** [comment]
I use jdk1.8

**Comment by thelastender** [comment]
When I call Fiber.unpark before Fiber.start, the result is the same. So, I think the way to fix this is forbid change fiber state(call unpark or some other method) when a fiber is not start.

**Comment by pron** [comment]
I am not sure I understand the problem. Like `LockSupport.park/unpark` and all other synchronization primitives, `park/unpark` use a lease system and `park` must be called inside a loop as "before" and "after" are not things you'd want to count on in concurrent scenarios. 

Here's why it must be so. Suppose thread (or fiber -- same thing) A wakes up thread B, blocks and waits for B to wake it up. Because A must start B before parking, there is absolutely no way of ensuring whether B will get to unpark A only after A parks. `park` must, therefore, allow for cases where its matching `unpark` is called before it. An `unpark` must therefore grant a lease, and `park` must only be called inside a loop.

In short, `park` and `unpark` are low-level concurrency primitives that should only be used when writing synchronization mechanisms.

**Comment by thelastender** [comment]
Well, I mean the output of the example is not what I expected. I think the code above will output like this:

```
2017-03-03 14:48:41.240 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 0
f in B
2017-03-03 14:48:41.247 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 0
2017-03-03 14:48:41.247 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 1
f in B
```

wait 3 seconds then output:

```
2017-03-03 14:48:44.248 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 1
2017-03-03 14:48:44.248 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 2
f in B
```

wait 3 seconds then output:

```
2017-03-03 14:48:44.248 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) after 2
2017-03-03 14:48:44.249 DEBUG (ForkJoinPool-1-worker-1) (com.freejoy.fibers.test.MyFiberTest$1) before 3
f in B
```

......  (I omit the remaining output)

But the actual output is what I have posted above. So I want to figure out the actual meaning of `park` and `unpark` and I want to known why these output happened.

**Comment by thelastender** [comment]
I add a breakpoint to the code at line `f.unpark()` to debug. When I wait few seconds at line `f.unpark()` and then continue. The output is actually what I expected. It wait 3 seconds and then output three line each loop.

**Comment by pron** [comment]
Does this happen when you use `sleep` instead of `park`?

While this behavior is not what you'd expect (and I may decide to fix it), I don't believe it is actually bug, as `park` is allowed to wake up spuriously, and is basically meaningless unless run in a loop that checks for the wakeup condition. Think of `park` as "may block" and of timed park as "may block up to the timeout specified". This is the only semantics that makes sense.

If it happens with `sleep` as well then that's a bug because `sleep` means "block for at least the timeout specified or throw an `InterruptedException`".

**Comment by thelastender** [comment]
Oh! I see what you mean. Could you please show me a whole correct usage of `park` and `unpark`. I think maybe these codes show how `park` use.

```
...
while(condition A is not OK)
{
        Fiber.park();
}
...
```

Am I right? Even if I'm right, I still want to see your complete example.
Thank you very much!

**Comment by pron** [comment]
Yes. See the [documentation for `park`/`unpark` in the JDK](http://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/LockSupport.html).

**Comment by thelastender** [comment]
When I use `sleep`, the output is what I expected. 
Thanks very much!
